File: 02_Scripts/01_Data_Processing/03_sample_inventory.R
################################################################################
# Task 1.3: Create Sample Inventory and Overlap Analysis
################################################################################
# This script creates a comprehensive inventory of samples across all datasets
# and analyzes overlap for multi-omics integration.
#
# Outputs:
# - Sample inventory table
# - Overlap analysis
# - Integration-ready sample list
#
# Author: Data Analysis Pipeline
# Date: 2025-10-02
# Project: AML Multi-Omics Integration
################################################################################

library(readr)
library(readxl)
library(dplyr)
library(tibble)

#' Extract sample IDs from a dataframe
#' @param df Data frame
#' @param dataset_name Name of dataset
#' @return List with sample_ids, id_column, id_format
extract_sample_ids <- function(df, dataset_name) {

  if (dataset_name == "expression") {
    # Sample IDs are column names (skip first 4 annotation columns)
    sample_ids <- colnames(df)[5:ncol(df)]
    sample_ids <- sample_ids[grepl("^BA", sample_ids)]
    id_column <- "column_names"
    id_format <- "BA####R"

  } else if (dataset_name == "drug_auc") {
    sample_ids <- unique(df$dbgap_rnaseq_sample[!is.na(df$dbgap_rnaseq_sample)])
    id_column <- "dbgap_rnaseq_sample"
    id_format <- "BA####R"

  } else if (dataset_name == "clinical") {
    if ("dbgap_rnaseq_sample" %in% colnames(df)) {
      sample_ids <- unique(df$dbgap_rnaseq_sample[!is.na(df$dbgap_rnaseq_sample)])
      id_column <- "dbgap_rnaseq_sample"
      id_format <- "BA####R"
    } else if ("dbgap_subject_id" %in% colnames(df)) {
      sample_ids <- unique(df$dbgap_subject_id[!is.na(df$dbgap_subject_id)])
      id_column <- "dbgap_subject_id"
      id_format <- "numeric"
    } else {
      sample_ids <- character(0)
      id_column <- "unknown"
      id_format <- "unknown"
    }

  } else if (dataset_name == "mutations") {
    if ("dbgap_sample_id" %in% colnames(df)) {
      sample_ids <- unique(df$dbgap_sample_id[!is.na(df$dbgap_sample_id)])
      id_column <- "dbgap_sample_id"
      id_format <- "BA####D"
    } else {
      sample_ids <- character(0)
      id_column <- "unknown"
      id_format <- "unknown"
    }

  } else if (dataset_name == "raw_inhibitor") {
    sample_ids <- unique(df$dbgap_rnaseq_sample[!is.na(df$dbgap_rnaseq_sample)])
    id_column <- "dbgap_rnaseq_sample"
    id_format <- "BA####R"

  } else {
    sample_ids <- character(0)
    id_column <- "unknown"
    id_format <- "unknown"
  }

  return(list(
    sample_ids = sample_ids,
    id_column = id_column,
    id_format = id_format
  ))
}

#' Load all datasets
#' @param data_dir Path to data directory
#' @return List of datasets
load_datasets <- function(data_dir) {

  cat("Loading datasets...\n")
  cat(strrep("=", 80), "\n")

  datasets <- list()

  # 1. Expression data
  cat("\n1. Loading expression data...\n")
  tryCatch({
    expr <- read_delim(file.path(data_dir, "beataml_expression.txt"),
                      delim = "\t", show_col_types = FALSE)
    datasets$expression <- expr
    cat(sprintf("   ✓ Loaded: %d x %d\n", nrow(expr), ncol(expr)))
  }, error = function(e) {
    cat("   ✗ Error:", e$message, "\n")
    datasets$expression <- NULL
  })

  # 2. Drug AUC data
  cat("\n2. Loading drug AUC data...\n")
  tryCatch({
    drug_auc <- read_delim(file.path(data_dir, "beataml_drug_auc.txt"),
                          delim = "\t", show_col_types = FALSE)
    datasets$drug_auc <- drug_auc
    cat(sprintf("   ✓ Loaded: %d x %d\n", nrow(drug_auc), ncol(drug_auc)))
  }, error = function(e) {
    cat("   ✗ Error:", e$message, "\n")
    datasets$drug_auc <- NULL
  })

  # 3. Clinical data
  cat("\n3. Loading clinical data...\n")
  tryCatch({
    clinical <- read_excel(file.path(data_dir, "beataml_clinical.xlsx"))
    datasets$clinical <- clinical
    cat(sprintf("   ✓ Loaded: %d x %d\n", nrow(clinical), ncol(clinical)))
  }, error = function(e) {
    cat("   ✗ Error:", e$message, "\n")
    datasets$clinical <- NULL
  })

  # 4. Mutations data
  cat("\n4. Loading mutations data...\n")
  tryCatch({
    mutations <- read_delim(file.path(data_dir, "beataml_mutations.txt"),
                           delim = "\t", show_col_types = FALSE)
    datasets$mutations <- mutations
    cat(sprintf("   ✓ Loaded: %d x %d\n", nrow(mutations), ncol(mutations)))
  }, error = function(e) {
    cat("   ✗ Error:", e$message, "\n")
    datasets$mutations <- NULL
  })

  # 5. Raw inhibitor data
  cat("\n5. Loading raw inhibitor data...\n")
  tryCatch({
    raw_inhib <- read_delim(file.path(data_dir, "beataml_raw_inhibitor.txt"),
                           delim = "\t", show_col_types = FALSE)
    datasets$raw_inhibitor <- raw_inhib
    cat(sprintf("   ✓ Loaded: %d x %d\n", nrow(raw_inhib), ncol(raw_inhib)))
  }, error = function(e) {
    cat("   ✗ Error:", e$message, "\n")
    datasets$raw_inhibitor <- NULL
  })

  cat("\n", strrep("=", 80), "\n")
  return(datasets)
}

#' Create sample inventory
#' @param datasets List of datasets
#' @return List with inventory dataframe and sample_sets
create_sample_inventory <- function(datasets) {

  cat("\nCreating sample inventory...\n")
  cat(strrep("=", 80), "\n")

  inventory <- data.frame(
    data_type = character(),
    n_samples = integer(),
    sample_id_column = character(),
    id_format = character(),
    date_analyzed = character(),
    stringsAsFactors = FALSE
  )

  sample_sets <- list()

  for (dataset_name in names(datasets)) {
    df <- datasets[[dataset_name]]

    if (is.null(df)) {
      cat(sprintf("\n✗ Skipping %s (not loaded)\n", dataset_name))
      next
    }

    cat(sprintf("\nProcessing %s...\n", dataset_name))

    # Extract sample IDs
    result <- extract_sample_ids(df, dataset_name)
    sample_ids <- result$sample_ids
    id_column <- result$id_column
    id_format <- result$id_format

    n_samples <- length(sample_ids)

    cat(sprintf("  ID Column: %s\n", id_column))
    cat(sprintf("  ID Format: %s\n", id_format))
    cat(sprintf("  N Samples: %s\n", format(n_samples, big.mark = ",")))

    # Store sample set
    sample_sets[[dataset_name]] <- sample_ids

    # Add to inventory
    inventory <- rbind(inventory, data.frame(
      data_type = dataset_name,
      n_samples = n_samples,
      sample_id_column = id_column,
      id_format = id_format,
      date_analyzed = format(Sys.Date(), "%Y-%m-%d"),
      stringsAsFactors = FALSE
    ))
  }

  cat("\n", strrep("=", 80), "\n")

  return(list(
    inventory = inventory,
    sample_sets = sample_sets
  ))
}

#' Analyze overlap between datasets
#' @param sample_sets List of sample sets
#' @return Core samples
analyze_overlap <- function(sample_sets) {

  cat("\nAnalyzing sample overlap...\n")
  cat(strrep("=", 80), "\n")

  # Separate RNA and DNA samples
  rna_samples <- list()
  dna_samples <- list()

  for (name in names(sample_sets)) {
    samples <- sample_sets[[name]]
    if (length(samples) > 0) {
      if (is.character(samples) && grepl("R", samples[1])) {
        rna_samples[[name]] <- samples
      } else if (is.character(samples) && grepl("D", samples[1])) {
        dna_samples[[name]] <- samples
      }
    }
  }

  cat("\nRNA-seq based datasets:\n")
  for (name in names(rna_samples)) {
    cat(sprintf("  %s: %d samples\n", name, length(rna_samples[[name]])))
  }

  cat("\nDNA-seq based datasets:\n")
  for (name in names(dna_samples)) {
    cat(sprintf("  %s: %d samples\n", name, length(dna_samples[[name]])))
  }

  # Calculate overlaps for RNA samples
  if (length(rna_samples) >= 2) {
    cat("\n", strrep("-", 80), "\n")
    cat("RNA Sample Overlaps:\n")
    cat(strrep("-", 80), "\n")

    names_list <- names(rna_samples)
    for (i in 1:(length(names_list)-1)) {
      for (j in (i+1):length(names_list)) {
        name1 <- names_list[i]
        name2 <- names_list[j]

        overlap <- intersect(rna_samples[[name1]], rna_samples[[name2]])
        only1 <- setdiff(rna_samples[[name1]], rna_samples[[name2]])
        only2 <- setdiff(rna_samples[[name2]], rna_samples[[name1]])

        cat(sprintf("\n%s ∩ %s:\n", name1, name2))
        cat(sprintf("  Overlap: %d samples\n", length(overlap)))
        cat(sprintf("  Only in %s: %d samples\n", name1, length(only1)))
        cat(sprintf("  Only in %s: %d samples\n", name2, length(only2)))
      }
    }
  }

  # Find core multi-omics samples
  core_samples <- character(0)

  if ("expression" %in% names(rna_samples) && "drug_auc" %in% names(rna_samples)) {
    core_samples <- intersect(rna_samples$expression, rna_samples$drug_auc)

    if ("clinical" %in% names(rna_samples)) {
      core_samples <- intersect(core_samples, rna_samples$clinical)
    }

    cat("\n", strrep("=", 80), "\n")
    cat("CORE MULTI-OMICS SAMPLES\n")
    cat(strrep("=", 80), "\n")
    cat(sprintf("Samples with Expression + Drug Response: %d\n", length(core_samples)))
  }

  return(core_samples)
}

#' Create integration table
#' @param datasets List of datasets
#' @param core_samples Vector of core sample IDs
#' @return Integration dataframe
create_integration_table <- function(datasets, core_samples) {

  cat("\nCreating integration table...\n")
  cat(strrep("=", 80), "\n")

  integration_data <- data.frame(
    sample_id = core_samples,
    stringsAsFactors = FALSE
  )

  # Get sample lists
  expr_samples <- extract_sample_ids(datasets$expression, "expression")$sample_ids
  drug_samples <- extract_sample_ids(datasets$drug_auc, "drug_auc")$sample_ids

  integration_data$has_expression <- integration_data$sample_id %in% expr_samples
  integration_data$has_drug_response <- integration_data$sample_id %in% drug_samples

  if (!is.null(datasets$clinical)) {
    clin_samples <- extract_sample_ids(datasets$clinical, "clinical")$sample_ids
    integration_data$has_clinical <- integration_data$sample_id %in% clin_samples
  } else {
    integration_data$has_clinical <- FALSE
  }

  # Check for mutations (map RNA -> DNA)
  if (!is.null(datasets$mutations)) {
    mut_samples <- extract_sample_ids(datasets$mutations, "mutations")$sample_ids
    integration_data$dna_sample_id <- gsub("R$", "D", integration_data$sample_id)
    integration_data$has_mutations <- integration_data$dna_sample_id %in% mut_samples
  } else {
    integration_data$has_mutations <- FALSE
  }

  cat(sprintf("\nIntegration table created: %d samples\n", nrow(integration_data)))

  cat("\nData availability:\n")
  for (col in c("has_expression", "has_drug_response", "has_clinical", "has_mutations")) {
    if (col %in% colnames(integration_data)) {
      n <- sum(integration_data[[col]], na.rm = TRUE)
      pct <- (n / nrow(integration_data)) * 100
      cat(sprintf("  %s: %d (%.1f%%)\n",
                  gsub("has_", "", col), n, pct))
    }
  }

  # Count complete cases
  complete <- integration_data[integration_data$has_expression &
                               integration_data$has_drug_response &
                               integration_data$has_mutations, ]
  cat(sprintf("\nComplete multi-omics (Expression + Drug + Mutations): %d samples\n",
              nrow(complete)))

  return(integration_data)
}

#' Main function
main <- function() {

  # Set paths
  script_dir <- dirname(sys.frame(1)$ofile)
  project_root <- dirname(dirname(script_dir))
  data_dir <- file.path(project_root, "01_Data", "BeatAML_Downloaded_Data")
  output_dir <- file.path(project_root, "03_Results", "01_Processed_Data")

  # Create output directory
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  cat(strrep("=", 80), "\n")
  cat("BeatAML Sample Inventory and Overlap Analysis\n")
  cat(strrep("=", 80), "\n")
  cat("Date:", format(Sys.time(), "%Y-%m-%d %H:%M:%S"), "\n")
  cat("Data Directory:", data_dir, "\n")
  cat("Output Directory:", output_dir, "\n")
  cat(strrep("=", 80), "\n")

  # Load datasets
  datasets <- load_datasets(data_dir)

  # Create inventory
  result <- create_sample_inventory(datasets)
  inventory_df <- result$inventory
  sample_sets <- result$sample_sets

  # Display inventory
  cat("\n", strrep("=", 80), "\n")
  cat("SAMPLE INVENTORY\n")
  cat(strrep("=", 80), "\n")
  print(inventory_df, row.names = FALSE)

  # Save inventory
  inventory_file <- file.path(output_dir, "sample_inventory.csv")
  write.csv(inventory_df, inventory_file, row.names = FALSE)
  cat(sprintf("\n✓ Saved inventory to: %s\n", inventory_file))

  # Analyze overlap
  core_samples <- analyze_overlap(sample_sets)

  # Create integration table
  if (length(core_samples) > 0) {
    integration_df <- create_integration_table(datasets, core_samples)

    # Save integration table
    integration_file <- file.path(output_dir, "sample_integration_table.csv")
    write.csv(integration_df, integration_file, row.names = FALSE)
    cat(sprintf("\n✓ Saved integration table to: %s\n", integration_file))

    # Save core sample list
    core_file <- file.path(output_dir, "core_multi_omics_samples.txt")
    writeLines(c(
      "# Core Multi-Omics Samples (Expression + Drug Response)",
      sprintf("# Generated: %s", format(Sys.time(), "%Y-%m-%d %H:%M:%S")),
      sprintf("# Total samples: %d", length(core_samples)),
      "",
      sort(core_samples)
    ), core_file)
    cat(sprintf("✓ Saved core sample list to: %s\n", core_file))
  }

  cat("\n", strrep("=", 80), "\n")
  cat("✓ Sample inventory analysis complete!\n")
  cat(strrep("=", 80), "\n")
}

# Run if not in interactive mode
if (!interactive()) {
  main()
} else {
  cat("Script loaded. Run main() to execute.\n")
}

File: 02_Scripts/01_Data_Processing/03_sample_inventory.py
"""
Task 1.3: Create Sample Inventory and Overlap Analysis
=======================================================
This script creates a comprehensive inventory of samples across all datasets
and analyzes overlap for multi-omics integration.

Outputs:
- Sample inventory table
- Overlap analysis (Venn-style)
- Integration-ready sample list

Author: Data Analysis Pipeline
Date: 2025-10-02
Project: AML Multi-Omics Integration
"""

import os
import sys
from pathlib import Path
from datetime import datetime
import pandas as pd
import numpy as np

# Set UTF-8 encoding for Windows console
if sys.platform == 'win32':
    import io
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')

def extract_sample_ids(df, dataset_name):
    """Extract sample IDs from a dataframe based on dataset type."""

    if dataset_name == "expression":
        # Sample IDs are column names (skip first 4 annotation columns)
        sample_ids = [col for col in df.columns[4:] if col.startswith('BA')]
        id_column = "column_names"
        id_format = "BA####R"

    elif dataset_name == "drug_auc":
        # Sample IDs in dbgap_rnaseq_sample column
        sample_ids = df['dbgap_rnaseq_sample'].dropna().unique().tolist()
        id_column = "dbgap_rnaseq_sample"
        id_format = "BA####R"

    elif dataset_name == "clinical":
        # Sample IDs might be in dbgap_rnaseq_sample or similar
        if 'dbgap_rnaseq_sample' in df.columns:
            sample_ids = df['dbgap_rnaseq_sample'].dropna().unique().tolist()
            id_column = "dbgap_rnaseq_sample"
            id_format = "BA####R"
        elif 'dbgap_subject_id' in df.columns:
            sample_ids = df['dbgap_subject_id'].dropna().unique().tolist()
            id_column = "dbgap_subject_id"
            id_format = "numeric"
        else:
            sample_ids = []
            id_column = "unknown"
            id_format = "unknown"

    elif dataset_name == "mutations":
        # Sample IDs in dbgap_sample_id column (DNA samples)
        if 'dbgap_sample_id' in df.columns:
            sample_ids = df['dbgap_sample_id'].dropna().unique().tolist()
            id_column = "dbgap_sample_id"
            id_format = "BA####D"
        else:
            sample_ids = []
            id_column = "unknown"
            id_format = "unknown"

    elif dataset_name == "raw_inhibitor":
        # Sample IDs in dbgap_rnaseq_sample column
        sample_ids = df['dbgap_rnaseq_sample'].dropna().unique().tolist()
        id_column = "dbgap_rnaseq_sample"
        id_format = "BA####R"

    else:
        sample_ids = []
        id_column = "unknown"
        id_format = "unknown"

    return sample_ids, id_column, id_format

def load_datasets(data_dir):
    """Load all datasets and extract sample information."""

    print("Loading datasets...")
    print("=" * 80)

    datasets = {}

    # 1. Expression data
    print("\n1. Loading expression data...")
    try:
        expr = pd.read_csv(data_dir / "beataml_expression.txt", sep='\t', low_memory=False)
        datasets['expression'] = expr
        print(f"   ✓ Loaded: {expr.shape}")
    except Exception as e:
        print(f"   ✗ Error: {e}")
        datasets['expression'] = None

    # 2. Drug AUC data
    print("\n2. Loading drug AUC data...")
    try:
        drug_auc = pd.read_csv(data_dir / "beataml_drug_auc.txt", sep='\t', low_memory=False)
        datasets['drug_auc'] = drug_auc
        print(f"   ✓ Loaded: {drug_auc.shape}")
    except Exception as e:
        print(f"   ✗ Error: {e}")
        datasets['drug_auc'] = None

    # 3. Clinical data
    print("\n3. Loading clinical data...")
    try:
        clinical = pd.read_excel(data_dir / "beataml_clinical.xlsx")
        datasets['clinical'] = clinical
        print(f"   ✓ Loaded: {clinical.shape}")
    except Exception as e:
        print(f"   ✗ Error: {e}")
        datasets['clinical'] = None

    # 4. Mutations data
    print("\n4. Loading mutations data...")
    try:
        mutations = pd.read_csv(data_dir / "beataml_mutations.txt", sep='\t', low_memory=False)
        datasets['mutations'] = mutations
        print(f"   ✓ Loaded: {mutations.shape}")
    except Exception as e:
        print(f"   ✗ Error: {e}")
        datasets['mutations'] = None

    # 5. Raw inhibitor data
    print("\n5. Loading raw inhibitor data...")
    try:
        raw_inhib = pd.read_csv(data_dir / "beataml_raw_inhibitor.txt", sep='\t', low_memory=False)
        datasets['raw_inhibitor'] = raw_inhib
        print(f"   ✓ Loaded: {raw_inhib.shape}")
    except Exception as e:
        print(f"   ✗ Error: {e}")
        datasets['raw_inhibitor'] = None

    print("\n" + "=" * 80)
    return datasets

def create_sample_inventory(datasets):
    """Create comprehensive sample inventory."""

    print("\nCreating sample inventory...")
    print("=" * 80)

    inventory = []
    sample_sets = {}

    for dataset_name, df in datasets.items():
        if df is None:
            print(f"\n✗ Skipping {dataset_name} (not loaded)")
            continue

        print(f"\nProcessing {dataset_name}...")

        # Extract sample IDs
        sample_ids, id_column, id_format = extract_sample_ids(df, dataset_name)

        n_samples = len(sample_ids)

        print(f"  ID Column: {id_column}")
        print(f"  ID Format: {id_format}")
        print(f"  N Samples: {n_samples:,}")

        # Store sample set for overlap analysis
        sample_sets[dataset_name] = set(sample_ids)

        # Add to inventory
        inventory.append({
            'data_type': dataset_name,
            'n_samples': n_samples,
            'sample_id_column': id_column,
            'id_format': id_format,
            'date_analyzed': datetime.now().strftime('%Y-%m-%d')
        })

    print("\n" + "=" * 80)

    return pd.DataFrame(inventory), sample_sets

def analyze_overlap(sample_sets):
    """Analyze overlap between datasets."""

    print("\nAnalyzing sample overlap...")
    print("=" * 80)

    # Get RNA-seq sample sets (expression, drug_auc, raw_inhibitor, clinical)
    rna_samples = {}
    dna_samples = {}

    for name, samples in sample_sets.items():
        # Check sample format
        if samples and len(samples) > 0:
            sample_list = list(samples)
            if isinstance(sample_list[0], str) and 'R' in sample_list[0]:
                rna_samples[name] = samples
            elif isinstance(sample_list[0], str) and 'D' in sample_list[0]:
                dna_samples[name] = samples
            else:
                # Numeric IDs - try to map
                pass

    print("\nRNA-seq based datasets:")
    for name, samples in rna_samples.items():
        print(f"  {name}: {len(samples)} samples")

    print("\nDNA-seq based datasets:")
    for name, samples in dna_samples.items():
        print(f"  {name}: {len(samples)} samples")

    # Calculate overlaps for RNA samples
    if len(rna_samples) >= 2:
        print("\n" + "-" * 80)
        print("RNA Sample Overlaps:")
        print("-" * 80)

        names = list(rna_samples.keys())
        for i, name1 in enumerate(names):
            for name2 in names[i+1:]:
                overlap = rna_samples[name1] & rna_samples[name2]
                only1 = rna_samples[name1] - rna_samples[name2]
                only2 = rna_samples[name2] - rna_samples[name1]

                print(f"\n{name1} ∩ {name2}:")
                print(f"  Overlap: {len(overlap)} samples")
                print(f"  Only in {name1}: {len(only1)} samples")
                print(f"  Only in {name2}: {len(only2)} samples")

    # Find core multi-omics samples (expression + drug + clinical)
    if 'expression' in rna_samples and 'drug_auc' in rna_samples:
        core_samples = rna_samples['expression'] & rna_samples['drug_auc']

        if 'clinical' in rna_samples:
            core_samples = core_samples & rna_samples['clinical']

        print("\n" + "=" * 80)
        print("CORE MULTI-OMICS SAMPLES")
        print("=" * 80)
        print(f"Samples with Expression + Drug Response: {len(core_samples)}")

        return core_samples

    return set()

def create_integration_table(datasets, core_samples):
    """Create integration-ready sample table."""

    print("\nCreating integration table...")
    print("=" * 80)

    integration_data = []

    for sample in sorted(core_samples):
        row = {'sample_id': sample}

        # Check presence in each dataset
        row['has_expression'] = sample in extract_sample_ids(datasets['expression'], 'expression')[0]
        row['has_drug_response'] = sample in extract_sample_ids(datasets['drug_auc'], 'drug_auc')[0]

        if datasets['clinical'] is not None:
            clinical_samples = extract_sample_ids(datasets['clinical'], 'clinical')[0]
            row['has_clinical'] = sample in clinical_samples
        else:
            row['has_clinical'] = False

        # Check for mutations (need to map RNA -> DNA ID)
        # BA####R -> BA####D
        if datasets['mutations'] is not None:
            dna_id = sample.replace('R', 'D')
            mutation_samples = extract_sample_ids(datasets['mutations'], 'mutations')[0]
            row['has_mutations'] = dna_id in mutation_samples
            if row['has_mutations']:
                row['dna_sample_id'] = dna_id
        else:
            row['has_mutations'] = False

        integration_data.append(row)

    df = pd.DataFrame(integration_data)

    print(f"\nIntegration table created: {len(df)} samples")
    print("\nData availability:")
    for col in ['has_expression', 'has_drug_response', 'has_clinical', 'has_mutations']:
        if col in df.columns:
            n = df[col].sum()
            pct = (n / len(df)) * 100
            print(f"  {col.replace('has_', '')}: {n} ({pct:.1f}%)")

    # Count complete cases
    complete_multi_omics = df[df[['has_expression', 'has_drug_response', 'has_mutations']].all(axis=1)]
    print(f"\nComplete multi-omics (Expression + Drug + Mutations): {len(complete_multi_omics)} samples")

    return df

def main():
    """Main function."""

    # Set paths
    project_root = Path(__file__).parent.parent.parent
    data_dir = project_root / "01_Data" / "BeatAML_Downloaded_Data"
    output_dir = project_root / "03_Results" / "01_Processed_Data"
    output_dir.mkdir(parents=True, exist_ok=True)

    print("=" * 80)
    print("BeatAML Sample Inventory and Overlap Analysis")
    print("=" * 80)
    print(f"Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print(f"Data Directory: {data_dir}")
    print(f"Output Directory: {output_dir}")
    print("=" * 80)

    # Load datasets
    datasets = load_datasets(data_dir)

    # Create sample inventory
    inventory_df, sample_sets = create_sample_inventory(datasets)

    # Display inventory
    print("\n" + "=" * 80)
    print("SAMPLE INVENTORY")
    print("=" * 80)
    print(inventory_df.to_string(index=False))

    # Save inventory
    inventory_file = output_dir / "sample_inventory.csv"
    inventory_df.to_csv(inventory_file, index=False)
    print(f"\n✓ Saved inventory to: {inventory_file}")

    # Analyze overlap
    core_samples = analyze_overlap(sample_sets)

    # Create integration table
    if len(core_samples) > 0:
        integration_df = create_integration_table(datasets, core_samples)

        # Save integration table
        integration_file = output_dir / "sample_integration_table.csv"
        integration_df.to_csv(integration_file, index=False)
        print(f"\n✓ Saved integration table to: {integration_file}")

        # Save core sample list
        core_file = output_dir / "core_multi_omics_samples.txt"
        with open(core_file, 'w') as f:
            f.write("# Core Multi-Omics Samples (Expression + Drug Response)\n")
            f.write(f"# Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
            f.write(f"# Total samples: {len(core_samples)}\n\n")
            for sample in sorted(core_samples):
                f.write(f"{sample}\n")
        print(f"✓ Saved core sample list to: {core_file}")

    print("\n" + "=" * 80)
    print("✓ Sample inventory analysis complete!")
    print("=" * 80)

if __name__ == "__main__":
    main()
